File: py_modules/sharelink.py
"""
Парсер share-ссылок (vless:// vmess:// ss:// trojan:// hysteria2://) и
base64-подписок v2ray-формата → готовый Clash/Mihomo YAML-конфиг.

Зачем: поле импорта раньше делало urllib.urlopen(url) и падало на
`unknown url type: vless`, потому что vless:// — это не подписка (HTTP-адрес
со списком нод), а одна нода, закодированная в строку. Здесь мы парсим такие
строки (и base64-списки из них) сами и собираем валидный конфиг.
"""

import base64
import binascii
import json
import io
import urllib.parse
from typing import List, Optional, Tuple

from ruamel.yaml import YAML

SCHEMES = ("vless://", "vmess://", "ss://", "trojan://",
           "hysteria2://", "hy2://")

_yaml = YAML()
_yaml.width = float("inf")
_yaml.default_flow_style = False


# ──────────────────────────── helpers ────────────────────────────

def _b64decode(s: str) -> Optional[str]:
    """Декод base64 (обычный и url-safe), терпимо к отсутствию паддинга."""
    s = s.strip()
    if not s:
        return None
    for fn in (base64.b64decode, base64.urlsafe_b64decode):
        try:
            pad = "=" * (-len(s) % 4)
            return fn(s + pad).decode("utf-8", "replace")
        except (binascii.Error, ValueError):
            continue
    return None


def _name_from_fragment(url: str, fallback: str) -> str:
    frag = urllib.parse.urlparse(url).fragment
    name = urllib.parse.unquote(frag).strip() if frag else ""
    return name or fallback


def _qs(url: str) -> dict:
    """query-параметры share-ссылки в плоский dict (последнее значение)."""
    q = urllib.parse.urlparse(url).query
    out = {}
    for k, v in urllib.parse.parse_qsl(q, keep_blank_values=True):
        out[k] = v
    return out


def _truthy(v) -> bool:
    return str(v).lower() in ("1", "true", "yes")


def _split_lines(text: str) -> List[str]:
    """Разбить на строки share-ссылки (одна или несколько / base64-список)."""
    text = text.strip()
    # base64-подписка целиком? — декодируем и работаем с результатом
    if not any(text.lower().startswith(s) for s in SCHEMES):
        dec = _b64decode(text)
        if dec and any(s in dec for s in SCHEMES):
            text = dec
    lines = []
    for raw in text.replace("\r", "\n").split("\n"):
        ln = raw.strip()
        if ln and any(ln.lower().startswith(s) for s in SCHEMES):
            lines.append(ln)
    return lines


def looks_like_sharelink(text: str) -> bool:
    """Похоже ли это на share-ссылку / base64-список таких ссылок."""
    return bool(_split_lines(text))


# ──────────────────────────── parsers ────────────────────────────

def _net_opts(proxy: dict, params: dict, network: str) -> None:
    """Заполнить ws/grpc/http транспорт по query-параметрам."""
    host = params.get("host", "")
    path = params.get("path", "")
    if network in ("ws", "httpupgrade"):
        ws = {}
        if path:
            ws["path"] = path
        if host:
            ws["headers"] = {"Host": host}
        if ws:
            proxy["ws-opts"] = ws
    elif network == "grpc":
        svc = params.get("serviceName") or params.get("servicename") or path
        if svc:
            proxy["grpc-opts"] = {"grpc-service-name": svc}
    elif network in ("http", "h2"):
        h2 = {}
        if path:
            h2["path"] = path
        if host:
            h2["host"] = [h for h in host.split(",") if h]
        if h2:
            proxy["h2-opts"] = h2


def _tls_opts(proxy: dict, params: dict, default_sni: str = "") -> None:
    security = params.get("security", "").lower()
    sni = params.get("sni") or params.get("peer") or default_sni
    if security in ("tls", "reality", "xtls"):
        proxy["tls"] = True
    if sni:
        proxy["servername"] = sni
    fp = params.get("fp")
    if fp:
        proxy["client-fingerprint"] = fp
    alpn = params.get("alpn")
    if alpn:
        proxy["alpn"] = [a for a in alpn.split(",") if a]
    if _truthy(params.get("allowInsecure")) or _truthy(params.get("insecure")):
        proxy["skip-cert-verify"] = True
    if security == "reality":
        ro = {}
        if params.get("pbk"):
            ro["public-key"] = params["pbk"]
        if params.get("sid"):
            ro["short-id"] = params["sid"]
        if ro:
            proxy["reality-opts"] = ro


def _parse_vless(url: str, idx: int) -> Optional[dict]:
    u = urllib.parse.urlparse(url)
    if not u.hostname or not u.port:
        return None
    p = _qs(url)
    network = (p.get("type") or "tcp").lower()
    proxy = {
        "name": _name_from_fragment(url, f"vless-{idx}"),
        "type": "vless",
        "server": u.hostname,
        "port": int(u.port),
        "uuid": urllib.parse.unquote(u.username or ""),
        "network": "ws" if network == "httpupgrade" else network,
        "udp": True,
    }
    flow = p.get("flow")
    if flow:
        proxy["flow"] = flow
    _tls_opts(proxy, p, default_sni=p.get("host", ""))
    _net_opts(proxy, p, network)
    return proxy


def _parse_vmess(url: str, idx: int) -> Optional[dict]:
    body = url[len("vmess://"):]
    dec = _b64decode(body)
    if not dec:
        return None
    try:
        cfg = json.loads(dec)
    except (json.JSONDecodeError, ValueError):
        return None
    server = cfg.get("add")
    port = cfg.get("port")
    if not server or not port:
        return None
    network = str(cfg.get("net", "tcp")).lower()
    proxy = {
        "name": str(cfg.get("ps") or f"vmess-{idx}").strip() or f"vmess-{idx}",
        "type": "vmess",
        "server": server,
        "port": int(port),
        "uuid": cfg.get("id", ""),
        "alterId": int(cfg.get("aid", 0) or 0),
        "cipher": cfg.get("scy") or "auto",
        "network": "ws" if network == "httpupgrade" else network,
        "udp": True,
    }
    if str(cfg.get("tls", "")).lower() in ("tls", "reality"):
        proxy["tls"] = True
        sni = cfg.get("sni") or cfg.get("host")
        if sni:
            proxy["servername"] = sni
        if cfg.get("alpn"):
            proxy["alpn"] = [a for a in str(cfg["alpn"]).split(",") if a]
        if cfg.get("fp"):
            proxy["client-fingerprint"] = cfg["fp"]
    params = {"host": cfg.get("host", ""), "path": cfg.get("path", ""),
              "serviceName": cfg.get("path", "")}
    _net_opts(proxy, params, network)
    return proxy


def _parse_ss(url: str, idx: int) -> Optional[dict]:
    name = _name_from_fragment(url, f"ss-{idx}")
    body = url[len("ss://"):]
    if "#" in body:
        body = body.split("#", 1)[0]
    query = ""
    if "?" in body:
        body, query = body.split("?", 1)
    # форматы: BASE64(method:pass)@host:port  |  BASE64(method:pass@host:port)
    method = password = host = None
    port = None
    if "@" in body:
        userinfo, hostport = body.rsplit("@", 1)
        creds = _b64decode(userinfo) or urllib.parse.unquote(userinfo)
        if ":" in creds:
            method, password = creds.split(":", 1)
        if ":" in hostport:
            host, port_s = hostport.rsplit(":", 1)
            port = int(port_s) if port_s.isdigit() else None
    else:
        dec = _b64decode(body)
        if dec and "@" in dec and ":" in dec:
            creds, hostport = dec.rsplit("@", 1)
            method, password = creds.split(":", 1)
            host, port_s = hostport.rsplit(":", 1)
            port = int(port_s) if port_s.isdigit() else None
    if not (method and host and port):
        return None
    proxy = {
        "name": name,
        "type": "ss",
        "server": host,
        "port": port,
        "cipher": method,
        "password": password or "",
        "udp": True,
    }
    # SIP003 plugin (obfs / v2ray-plugin) — по возможности
    if query:
        qp = dict(urllib.parse.parse_qsl(query, keep_blank_values=True))
        plug = qp.get("plugin", "")
        if plug:
            parts = plug.split(";")
            pname = parts[0]
            opts = dict(o.split("=", 1) for o in parts[1:] if "=" in o)
            if pname in ("obfs-local", "simple-obfs"):
                proxy["plugin"] = "obfs"
                proxy["plugin-opts"] = {
                    "mode": opts.get("obfs", "http"),
                    "host": opts.get("obfs-host", ""),
                }
            elif pname == "v2ray-plugin":
                proxy["plugin"] = "v2ray-plugin"
                proxy["plugin-opts"] = {
                    "mode": opts.get("mode", "websocket"),
                    "host": opts.get("host", ""),
                    "path": opts.get("path", "/"),
                    "tls": "tls" in opts,
                }
    return proxy


def _parse_trojan(url: str, idx: int) -> Optional[dict]:
    u = urllib.parse.urlparse(url)
    if not u.hostname or not u.port:
        return None
    p = _qs(url)
    network = (p.get("type") or "tcp").lower()
    proxy = {
        "name": _name_from_fragment(url, f"trojan-{idx}"),
        "type": "trojan",
        "server": u.hostname,
        "port": int(u.port),
        "password": urllib.parse.unquote(u.username or ""),
        "udp": True,
    }
    sni = p.get("sni") or p.get("peer")
    if sni:
        proxy["sni"] = sni
    if p.get("alpn"):
        proxy["alpn"] = [a for a in p["alpn"].split(",") if a]
    if p.get("fp"):
        proxy["client-fingerprint"] = p["fp"]
    if _truthy(p.get("allowInsecure")) or _truthy(p.get("insecure")):
        proxy["skip-cert-verify"] = True
    if network in ("ws", "grpc", "http", "h2"):
        proxy["network"] = "ws" if network == "httpupgrade" else network
        _net_opts(proxy, p, network)
    return proxy


def _parse_hysteria2(url: str, idx: int) -> Optional[dict]:
    u = urllib.parse.urlparse(url)
    if not u.hostname or not u.port:
        return None
    p = _qs(url)
    proxy = {
        "name": _name_from_fragment(url, f"hy2-{idx}"),
        "type": "hysteria2",
        "server": u.hostname,
        "port": int(u.port),
        "password": urllib.parse.unquote(u.username or u.password or ""),
    }
    sni = p.get("sni") or p.get("peer")
    if sni:
        proxy["sni"] = sni
    if _truthy(p.get("insecure")) or _truthy(p.get("allowInsecure")):
        proxy["skip-cert-verify"] = True
    if p.get("obfs"):
        proxy["obfs"] = p["obfs"]
        if p.get("obfs-password"):
            proxy["obfs-password"] = p["obfs-password"]
    return proxy


_DISPATCH = (
    ("vless://", _parse_vless),
    ("vmess://", _parse_vmess),
    ("ss://", _parse_ss),
    ("trojan://", _parse_trojan),
    ("hysteria2://", _parse_hysteria2),
    ("hy2://", _parse_hysteria2),
)


def _parse_one(line: str, idx: int) -> Optional[dict]:
    for scheme, fn in _DISPATCH:
        if line.lower().startswith(scheme):
            try:
                return fn(line, idx)
            except Exception:
                return None
    return None


def _dedup_names(proxies: List[dict]) -> None:
    seen = {}
    for pr in proxies:
        base = pr["name"]
        if base in seen:
            seen[base] += 1
            pr["name"] = f"{base} #{seen[base]}"
        else:
            seen[base] = 1


# ──────────────────────────── public API ────────────────────────────

def parse(text: str) -> Tuple[List[dict], str]:
    """
    Разобрать текст (одна ссылка / много / base64-список) в список clash-нод.
    Returns: (proxies, suggested_name)
    """
    lines = _split_lines(text)
    proxies = []
    for i, ln in enumerate(lines, 1):
        pr = _parse_one(ln, i)
        if pr:
            proxies.append(pr)
    _dedup_names(proxies)
    if not proxies:
        name = "sharelink"
    elif len(proxies) == 1:
        name = proxies[0]["name"]
    else:
        name = proxies[0]["name"] + f" +{len(proxies) - 1}"
    return proxies, name


def build_yaml(proxies: List[dict]) -> bytes:
    """Собрать валидный минимальный Clash-конфиг из списка нод."""
    names = [p["name"] for p in proxies]
    config = {
        "mixed-port": 7890,
        "mode": "rule",
        "proxies": proxies,
        "proxy-groups": [
            {
                "name": "PROXY",
                "type": "select",
                "proxies": names + ["DIRECT"],
            },
            {
                "name": "AUTO",
                "type": "url-test",
                "url": "http://www.gstatic.com/generate_204",
                "interval": 300,
                "lazy": True,
                "proxies": names or ["DIRECT"],
            },
        ],
        "rules": [
            "MATCH,PROXY",
        ],
    }
    buf = io.BytesIO()
    _yaml.dump(config, buf)
    return buf.getvalue()

File: desktop/internal/sharelink/sharelink.go
// Package sharelink — порт py_modules/sharelink.py на Go.
// Парсит share-ссылки (vless:// vmess:// ss:// trojan:// hysteria2:// hy2://)
// и base64-подписки v2ray-формата в список Clash/Mihomo-нод и собирает
// минимальный валидный конфиг.
package sharelink

import (
	"encoding/base64"
	"encoding/json"
	"net/url"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
)

var schemes = []string{"vless://", "vmess://", "ss://", "trojan://", "hysteria2://", "hy2://"}

// Proxy — clash-нода (произвольные поля; порядок ключей для mihomo неважен).
type Proxy = map[string]any

// ───────────── helpers ─────────────

func b64decode(s string) (string, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", false
	}
	pad := s
	if m := len(s) % 4; m != 0 {
		pad = s + strings.Repeat("=", 4-m)
	}
	for _, enc := range []*base64.Encoding{base64.StdEncoding, base64.URLEncoding} {
		if b, err := enc.DecodeString(pad); err == nil {
			return string(b), true
		}
	}
	for _, enc := range []*base64.Encoding{base64.RawStdEncoding, base64.RawURLEncoding} {
		if b, err := enc.DecodeString(strings.TrimRight(s, "=")); err == nil {
			return string(b), true
		}
	}
	return "", false
}

func nameFromFragment(u *url.URL, fallback string) string {
	if n := strings.TrimSpace(u.Fragment); n != "" {
		return n
	}
	return fallback
}

func qs(u *url.URL) map[string]string {
	out := map[string]string{}
	for k, vs := range u.Query() {
		if len(vs) > 0 {
			out[k] = vs[len(vs)-1]
		}
	}
	return out
}

func truthy(v string) bool {
	switch strings.ToLower(v) {
	case "1", "true", "yes":
		return true
	}
	return false
}

func splitCSV(s string) []any {
	var out []any
	for _, p := range strings.Split(s, ",") {
		if p != "" {
			out = append(out, p)
		}
	}
	return out
}

func hasSchemePrefix(s string) bool {
	low := strings.ToLower(s)
	for _, sc := range schemes {
		if strings.HasPrefix(low, sc) {
			return true
		}
	}
	return false
}

func splitLines(text string) []string {
	text = strings.TrimSpace(text)
	if !hasSchemePrefix(text) {
		if dec, ok := b64decode(text); ok {
			low := strings.ToLower(dec)
			for _, sc := range schemes {
				if strings.Contains(low, sc) {
					text = dec
					break
				}
			}
		}
	}
	var lines []string
	for _, raw := range strings.Split(strings.ReplaceAll(text, "\r", "\n"), "\n") {
		ln := strings.TrimSpace(raw)
		if ln != "" && hasSchemePrefix(ln) {
			lines = append(lines, ln)
		}
	}
	return lines
}

// LooksLikeSharelink — похоже ли на share-ссылку / base64-список.
func LooksLikeSharelink(text string) bool {
	return len(splitLines(text)) > 0
}

// ───────────── транспорт / tls ─────────────

func netOpts(p Proxy, params map[string]string, network string) {
	host, path := params["host"], params["path"]
	switch network {
	case "ws", "httpupgrade":
		ws := Proxy{}
		if path != "" {
			ws["path"] = path
		}
		if host != "" {
			ws["headers"] = Proxy{"Host": host}
		}
		if len(ws) > 0 {
			p["ws-opts"] = ws
		}
	case "grpc":
		svc := params["serviceName"]
		if svc == "" {
			svc = params["servicename"]
		}
		if svc == "" {
			svc = path
		}
		if svc != "" {
			p["grpc-opts"] = Proxy{"grpc-service-name": svc}
		}
	case "http", "h2":
		h2 := Proxy{}
		if path != "" {
			h2["path"] = path
		}
		if host != "" {
			h2["host"] = splitCSV(host)
		}
		if len(h2) > 0 {
			p["h2-opts"] = h2
		}
	}
}

func tlsOpts(p Proxy, params map[string]string, defaultSNI string) {
	security := strings.ToLower(params["security"])
	sni := params["sni"]
	if sni == "" {
		sni = params["peer"]
	}
	if sni == "" {
		sni = defaultSNI
	}
	if security == "tls" || security == "reality" || security == "xtls" {
		p["tls"] = true
	}
	if sni != "" {
		p["servername"] = sni
	}
	if fp := params["fp"]; fp != "" {
		p["client-fingerprint"] = fp
	}
	if alpn := params["alpn"]; alpn != "" {
		p["alpn"] = splitCSV(alpn)
	}
	if truthy(params["allowInsecure"]) || truthy(params["insecure"]) {
		p["skip-cert-verify"] = true
	}
	if security == "reality" {
		ro := Proxy{}
		if v := params["pbk"]; v != "" {
			ro["public-key"] = v
		}
		if v := params["sid"]; v != "" {
			ro["short-id"] = v
		}
		if len(ro) > 0 {
			p["reality-opts"] = ro
		}
	}
}

// ───────────── парсеры ─────────────

func parseVless(raw string, idx int) Proxy {
	u, err := url.Parse(raw)
	if err != nil || u.Hostname() == "" || u.Port() == "" {
		return nil
	}
	port, err := strconv.Atoi(u.Port())
	if err != nil {
		return nil
	}
	p := qs(u)
	network := strings.ToLower(p["type"])
	if network == "" {
		network = "tcp"
	}
	net := network
	if net == "httpupgrade" {
		net = "ws"
	}
	uuid := ""
	if u.User != nil {
		uuid = u.User.Username()
	}
	proxy := Proxy{
		"name": nameFromFragment(u, "vless-"+strconv.Itoa(idx)),
		"type": "vless", "server": u.Hostname(), "port": port,
		"uuid": uuid, "network": net, "udp": true,
	}
	if flow := p["flow"]; flow != "" {
		proxy["flow"] = flow
	}
	tlsOpts(proxy, p, p["host"])
	netOpts(proxy, p, network)
	return proxy
}

func parseVmess(raw string, idx int) Proxy {
	dec, ok := b64decode(strings.TrimPrefix(raw, "vmess://"))
	if !ok {
		return nil
	}
	var cfg map[string]any
	if err := json.Unmarshal([]byte(dec), &cfg); err != nil {
		return nil
	}
	server := asStr(cfg["add"])
	port := asInt(cfg["port"])
	if server == "" || port == 0 {
		return nil
	}
	network := strings.ToLower(asStr(cfg["net"]))
	if network == "" {
		network = "tcp"
	}
	net := network
	if net == "httpupgrade" {
		net = "ws"
	}
	name := strings.TrimSpace(asStr(cfg["ps"]))
	if name == "" {
		name = "vmess-" + strconv.Itoa(idx)
	}
	proxy := Proxy{
		"name": name, "type": "vmess", "server": server, "port": port,
		"uuid": asStr(cfg["id"]), "alterId": asInt(cfg["aid"]),
		"cipher": orStr(asStr(cfg["scy"]), "auto"), "network": net, "udp": true,
	}
	if t := strings.ToLower(asStr(cfg["tls"])); t == "tls" || t == "reality" {
		proxy["tls"] = true
		sni := orStr(asStr(cfg["sni"]), asStr(cfg["host"]))
		if sni != "" {
			proxy["servername"] = sni
		}
		if a := asStr(cfg["alpn"]); a != "" {
			proxy["alpn"] = splitCSV(a)
		}
		if fp := asStr(cfg["fp"]); fp != "" {
			proxy["client-fingerprint"] = fp
		}
	}
	netOpts(proxy, map[string]string{"host": asStr(cfg["host"]), "path": asStr(cfg["path"]), "serviceName": asStr(cfg["path"])}, network)
	return proxy
}

func parseSS(raw string, idx int) Proxy {
	u0, _ := url.Parse(raw)
	name := "ss-" + strconv.Itoa(idx)
	if u0 != nil {
		name = nameFromFragment(u0, name)
	}
	body := strings.TrimPrefix(raw, "ss://")
	if i := strings.Index(body, "#"); i >= 0 {
		body = body[:i]
	}
	query := ""
	if i := strings.Index(body, "?"); i >= 0 {
		query = body[i+1:]
		body = body[:i]
	}
	var method, password, host string
	var port int
	if i := strings.LastIndex(body, "@"); i >= 0 {
		userinfo, hostport := body[:i], body[i+1:]
		creds, ok := b64decode(userinfo)
		if !ok {
			creds, _ = url.QueryUnescape(userinfo)
		}
		if j := strings.Index(creds, ":"); j >= 0 {
			method, password = creds[:j], creds[j+1:]
		}
		if j := strings.LastIndex(hostport, ":"); j >= 0 {
			host = hostport[:j]
			port, _ = strconv.Atoi(hostport[j+1:])
		}
	} else if dec, ok := b64decode(body); ok && strings.Contains(dec, "@") && strings.Contains(dec, ":") {
		i := strings.LastIndex(dec, "@")
		creds, hostport := dec[:i], dec[i+1:]
		if j := strings.Index(creds, ":"); j >= 0 {
			method, password = creds[:j], creds[j+1:]
		}
		if j := strings.LastIndex(hostport, ":"); j >= 0 {
			host = hostport[:j]
			port, _ = strconv.Atoi(hostport[j+1:])
		}
	}
	if method == "" || host == "" || port == 0 {
		return nil
	}
	proxy := Proxy{
		"name": name, "type": "ss", "server": host, "port": port,
		"cipher": method, "password": password, "udp": true,
	}
	if query != "" {
		qp, _ := url.ParseQuery(query)
		if plug := qp.Get("plugin"); plug != "" {
			parts := strings.Split(plug, ";")
			pname := parts[0]
			opts := map[string]string{}
			for _, o := range parts[1:] {
				if j := strings.Index(o, "="); j >= 0 {
					opts[o[:j]] = o[j+1:]
				}
			}
			switch pname {
			case "obfs-local", "simple-obfs":
				proxy["plugin"] = "obfs"
				proxy["plugin-opts"] = Proxy{"mode": orStr(opts["obfs"], "http"), "host": opts["obfs-host"]}
			case "v2ray-plugin":
				_, tls := opts["tls"]
				proxy["plugin"] = "v2ray-plugin"
				proxy["plugin-opts"] = Proxy{"mode": orStr(opts["mode"], "websocket"), "host": opts["host"], "path": orStr(opts["path"], "/"), "tls": tls}
			}
		}
	}
	return proxy
}

func parseTrojan(raw string, idx int) Proxy {
	u, err := url.Parse(raw)
	if err != nil || u.Hostname() == "" || u.Port() == "" {
		return nil
	}
	port, err := strconv.Atoi(u.Port())
	if err != nil {
		return nil
	}
	p := qs(u)
	network := strings.ToLower(p["type"])
	if network == "" {
		network = "tcp"
	}
	pass := ""
	if u.User != nil {
		pass = u.User.Username()
	}
	proxy := Proxy{
		"name": nameFromFragment(u, "trojan-"+strconv.Itoa(idx)),
		"type": "trojan", "server": u.Hostname(), "port": port,
		"password": pass, "udp": true,
	}
	sni := orStr(p["sni"], p["peer"])
	if sni != "" {
		proxy["sni"] = sni
	}
	if a := p["alpn"]; a != "" {
		proxy["alpn"] = splitCSV(a)
	}
	if fp := p["fp"]; fp != "" {
		proxy["client-fingerprint"] = fp
	}
	if truthy(p["allowInsecure"]) || truthy(p["insecure"]) {
		proxy["skip-cert-verify"] = true
	}
	switch network {
	case "ws", "grpc", "http", "h2", "httpupgrade":
		net := network
		if net == "httpupgrade" {
			net = "ws"
		}
		proxy["network"] = net
		netOpts(proxy, p, network)
	}
	return proxy
}

func parseHysteria2(raw string, idx int) Proxy {
	u, err := url.Parse(raw)
	if err != nil || u.Hostname() == "" || u.Port() == "" {
		return nil
	}
	port, err := strconv.Atoi(u.Port())
	if err != nil {
		return nil
	}
	p := qs(u)
	pass := ""
	if u.User != nil {
		pass = u.User.Username()
		if pw, has := u.User.Password(); pass == "" && has {
			pass = pw
		}
	}
	proxy := Proxy{
		"name": nameFromFragment(u, "hy2-"+strconv.Itoa(idx)),
		"type": "hysteria2", "server": u.Hostname(), "port": port, "password": pass,
	}
	sni := orStr(p["sni"], p["peer"])
	if sni != "" {
		proxy["sni"] = sni
	}
	if truthy(p["insecure"]) || truthy(p["allowInsecure"]) {
		proxy["skip-cert-verify"] = true
	}
	if obfs := p["obfs"]; obfs != "" {
		proxy["obfs"] = obfs
		if op := p["obfs-password"]; op != "" {
			proxy["obfs-password"] = op
		}
	}
	return proxy
}

func parseOne(line string, idx int) Proxy {
	low := strings.ToLower(line)
	switch {
	case strings.HasPrefix(low, "vless://"):
		return parseVless(line, idx)
	case strings.HasPrefix(low, "vmess://"):
		return parseVmess(line, idx)
	case strings.HasPrefix(low, "ss://"):
		return parseSS(line, idx)
	case strings.HasPrefix(low, "trojan://"):
		return parseTrojan(line, idx)
	case strings.HasPrefix(low, "hysteria2://"), strings.HasPrefix(low, "hy2://"):
		return parseHysteria2(line, idx)
	}
	return nil
}

func dedupNames(proxies []Proxy) {
	seen := map[string]int{}
	for _, pr := range proxies {
		base := asStr(pr["name"])
		if seen[base] > 0 {
			seen[base]++
			pr["name"] = base + " #" + strconv.Itoa(seen[base])
		} else {
			seen[base] = 1
		}
	}
}

// Parse — разобрать текст в список нод + предложенное имя.
func Parse(text string) ([]Proxy, string) {
	lines := splitLines(text)
	var proxies []Proxy
	for i, ln := range lines {
		if pr := parseOne(ln, i+1); pr != nil {
			proxies = append(proxies, pr)
		}
	}
	dedupNames(proxies)
	name := "sharelink"
	if len(proxies) == 1 {
		name = asStr(proxies[0]["name"])
	} else if len(proxies) > 1 {
		name = asStr(proxies[0]["name"]) + " +" + strconv.Itoa(len(proxies)-1)
	}
	return proxies, name
}

// BuildYAML — собрать минимальный Clash-конфиг из списка нод.
func BuildYAML(proxies []Proxy) ([]byte, error) {
	names := make([]any, 0, len(proxies))
	for _, p := range proxies {
		names = append(names, asStr(p["name"]))
	}
	selectProxies := append(append([]any{}, names...), "DIRECT")
	autoProxies := names
	if len(autoProxies) == 0 {
		autoProxies = []any{"DIRECT"}
	}
	cfg := Proxy{
		"mixed-port": 7890,
		"mode":       "rule",
		"proxies":    toAnySlice(proxies),
		"proxy-groups": []any{
			Proxy{"name": "PROXY", "type": "select", "proxies": selectProxies},
			Proxy{"name": "AUTO", "type": "url-test", "url": "http://www.gstatic.com/generate_204", "interval": 300, "lazy": true, "proxies": autoProxies},
		},
		"rules": []any{"MATCH,PROXY"},
	}
	return yaml.Marshal(cfg)
}

// ───────────── мелкие приведения типов ─────────────

func asStr(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case float64:
		return strconv.FormatFloat(t, 'f', -1, 64)
	case int:
		return strconv.Itoa(t)
	case nil:
		return ""
	default:
		return ""
	}
}

func asInt(v any) int {
	switch t := v.(type) {
	case float64:
		return int(t)
	case int:
		return t
	case string:
		n, _ := strconv.Atoi(strings.TrimSpace(t))
		return n
	}
	return 0
}

func orStr(a, b string) string {
	if a != "" {
		return a
	}
	return b
}

func toAnySlice(ps []Proxy) []any {
	out := make([]any, len(ps))
	for i, p := range ps {
		out[i] = p
	}
	return out
}
